modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+
+For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:
+
+----
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields use `?`. For example, the following type allows the age field to be present or not:
+
+----
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+==== Close and Open Objects
+
+Close object means that the type only accepts an object value if there are no additional fields in the object,
+just those key values specified by the type.
+
+Open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accept objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+----
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a close object is `{| |}`. The following example, makes our user type only support the keys `firstName` and `lastName`, so this throws an exception because the `age` field is not accepted.
+
+----
+var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Function Type
+
+DataWeave is a functional language and functions are a first class citizen which means that functions are values
+and values have an associated type.
+
+The syntax to define a function type is:
+
+----
+(paramType: ParamType,...) -> ReturnType
+----
+
+For example, if you want to define a function that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:
+
+----
+(paramA: String, paramB: Number) -> Boolean
+----
+
+For example, you could define a function that receives another function as an argument:
+
+----
+fun findUserIndex(arr: Array<User>, condition: (userInArray: User, targetUser: User) -> Boolean): User = ???
+----
+
+If we call the function `findUserIndex` with some function that does not match the `condition` constraint `(User, User) -> Boolean`, DataWeave throws a type checking error.
+
+
+== Complex Types
+
+=== Any and Nothing
+
+In some cases you can not enforce a restriction as it accepts all the values. +
+
+* `Any` type accepts all possible values.
+* `Nothing` type accepts no value but it can be assigned to all the types. This type is not frequently used explicitly but it is by the type inference algorithm.
+
+=== Union Type
+
+Union type is used to compose types. The syntax to define a union type is:
+
+----
+TypeA | TypeB | ...
+----
+
+The following example defines a type that accepts any String and any Number:
+
+----
+var age: String | Number = ???
+----
```

Comment:
```suggestion
The following example defines a variable with a constraint that accepts a value of type String or Number:

----
var age: String | Number = if (payload.allStrings) "32" else 32
----
```
This could be a good example